src/rust/cryptography-x509-validation/src/policy/extension.rs

```diff
@@ -171,6 +162,180 @@ impl<B: CryptoOps> ExtensionPolicy<B> {
+        _cert: &Certificate<'_>,
+        extn: &Extension<'_>,
+    ) -> Result<(), PolicyError> {
+        let mut ekus: ExtendedKeyUsage<'_> = extn.value()?;
+
+        if ekus.any(|eku| eku == policy.extended_key_usage) {
+            Ok(())
+        } else {
+            Err(PolicyError::Other("required EKU not found"))
+        }
+    }
+}
+
+pub(crate) mod ca {
+    use cryptography_x509::{
+        certificate::Certificate,
+        extensions::{AuthorityKeyIdentifier, BasicConstraints, Extension, KeyUsage},
+    };
+
+    use crate::{
+        certificate::cert_is_self_signed,
+        ops::CryptoOps,
+        policy::{Policy, PolicyError},
+    };
+
+    pub(crate) fn authority_key_identifier<B: CryptoOps>(
+        policy: &Policy<'_, B>,
+        cert: &Certificate<'_>,
+        extn: Option<&Extension<'_>>,
+    ) -> Result<(), PolicyError> {
+        // The Authority Key Identifier MUST be present, with one exception:
+        // self-signed CAs may omit it.
+        match extn {
+            Some(extn) => {
+                let aki: AuthorityKeyIdentifier<'_> = extn.value()?;
+                // 7.1.2.11.1 Authority Key Identifier:
+                // authorityCertIssuer and authorityCertSerialNumber MUST NOT be present.
+                if aki.authority_cert_issuer.is_some() {
+                    return Err(
+                        "authorityKeyIdentifier must not contain authorityCertIssuer".into(),
+                    );
+                }
+
+                if aki.authority_cert_serial_number.is_some() {
+                    return Err(
+                        "authorityKeyIdentifier must not contain authorityCertSerialNumber".into(),
+                    );
+                }
+            }
+            None => {
+                if !cert_is_self_signed(cert, &policy.ops) {
+                    return Err(
+                        "authorityKeyIdentifier must be present in cross-signed CA certificate"
+                            .into(),
+                    );
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn key_usage<B: CryptoOps>(
+        _policy: &Policy<'_, B>,
+        _cert: &Certificate<'_>,
+        extn: &Extension<'_>,
+    ) -> Result<(), PolicyError> {
+        let key_usage: KeyUsage<'_> = extn.value()?;
+
+        if !key_usage.key_cert_sign() {
+            return Err("keyUsage.keyCertSign must be asserted in a CA certificate".into());
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn basic_constraints<B: CryptoOps>(
+        _policy: &Policy<'_, B>,
+        _cert: &Certificate<'_>,
+        extn: &Extension<'_>,
+    ) -> Result<(), PolicyError> {
+        let basic_constraints: BasicConstraints = extn.value()?;
+
+        if !basic_constraints.ca {
+            return Err("basicConstraints.cA must be asserted in a CA certificate".into());
+        }
+
+        Ok(())
+    }
+
+    // TODO: Validate EKUs for non-root CAs as well.
```

Comment:
TODO :-)

Reply:
Yeah, I think this should become an open issue (maybe part of a larger tracking meta-issue?) -- none of the implementations in the current limbo harness suite perform this validation, so doing it here would make this implementation stricter than others.

Reply:
Tracking here: https://github.com/pyca/cryptography/issues/2381#issuecomment-1816822268